org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinker.java

```diff
@@ -208,11 +201,17 @@ protected EObject createProxy(EObject obj, INode node, EReference eRef) {
 		return proxy;
 	}
 
+	private void ensureCacheInitialized() {
+		if (instantiableSubTypes == null) {
+			 instantiableSubTypes = new SimpleCache<EClass, EClass>(this::findInstantiableCompatible);
+		} 
+	}
 	/**
 	 * @since 2.7
 	 */
 	protected EClass getProxyType(EObject obj, EReference eRef) {
 		EClass referenceType = ecoreGenericsUtil.getReferenceType(eRef, obj.eClass());
+		ensureCacheInitialized();
```

Comment:
Maybe check first whether `referenceType` is abstract or an interface and if not, use it directly?

Reply:
Thanks, that is a good idea. I have adapted the commit accordingly.